### PR TITLE
Fix coverity issues

### PR DESCRIPTION
**Pull request by stiepan**
## Category:
**Other** (*e.g. Documentation, Tests, Configuration*)

## Description:
* Error checking and dynamic casting issues
* Adjusted the npy header parser: 
   * check afeter dyn_casting of file handle
   * make sure to allocate the extra byte for 10 bytes token for the `\0` trick
   * make sure to include the 10 bytes offset in the size of the read for the header
   * adjust the comment about the possible header size

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3472

### CODE REVIEW

File: dali/pipeline/pipeline.cc [hunk 1]
```diff
@@ -162,15 +162,30 @@ Pipeline::~Pipeline() {
...
+        mixed_op_ptr->BreakWaiting();
+      } break;
+      case OpType::GPU: {
+        auto *gpu_op_ptr = dynamic_cast<InputOperator<GPUBackend> *>(op_ptr);
+        if (!gpu_op_ptr) {
+          std::cerr << "Failed to break waiting - expected GPU op" << std::endl;
+          std::terminate();
+        }
+        gpu_op_ptr->BreakWaiting();
+      } break;
```
Comment (stiepan) [inline dali/pipeline/pipeline.cc:188]: Not entirely sure about it, but maybe it is better to fail in more controlled manner here.

File: dali/util/numpy.cc [hunk 2]
```diff
@@ -183,43 +183,48 @@ void ParseHeaderItself(HeaderData &parsed_header, char *data, size_t header_len)
...
+  size_t token_len = 10;
+  size_t token_read_len = align_up(token_len + 1, o_direct_read_len_alignm);
+  auto token_mem =
+      mm::alloc_raw_shared<char, mm::memory_kind::host>(token_read_len, o_direct_alignm);
   char *token = token_mem.get();
   auto file = dynamic_cast<ODirectFileStream*>(src);
-  int64_t nread = file->ReadAt(token, aligned_len, 0);
-  DALI_ENFORCE(nread <= static_cast<Index>(aligned_len) &&
-               nread >= static_cast<Index>(std::min(src->Size(), aligned_len)),
+  DALI_ENFORCE(
```
Comment (stiepan) [inline dali/util/numpy.cc:192]: Check for nullptr.

File: dali/util/numpy.cc [hunk 3]
```diff
@@ -183,43 +183,48 @@ void ParseHeaderItself(HeaderData &parsed_header, char *data, size_t header_len)
...
   auto header_len = GetHeaderLen(token);
 
   int64 offset = 6 + 1 + 1 + 2;
   // the header_len can be 4GiB according to the NPYv2 file format
   // specification: https://numpy.org/neps/nep-0001-npy-format.html
   // while this allocation could be sizable, it is performed on the host.
-  auto block_start = align_down(offset, o_direct_alignm);
-  auto block_end = align_up(header_len + 1, o_direct_alignm);
-  aligned_len = align_up(block_end - block_start, o_direct_read_len_alignm);
+  size_t aligned_header_len =
```
Comment (stiepan) [inline dali/util/numpy.cc:210]: We read from 0 offset, so take the whole header + token length not just header_len `file->ReadAt(token_mem.get(), aligned_header_len, 0);`

File: dali/util/numpy.cc [hunk 4]
```diff
@@ -183,43 +183,48 @@ void ParseHeaderItself(HeaderData &parsed_header, char *data, size_t header_len)
...
+  auto char_tmp = token[token_len];
+  token[token_len] = '\0';
 
   auto header_len = GetHeaderLen(token);
 
   int64 offset = 6 + 1 + 1 + 2;
   // the header_len can be 4GiB according to the NPYv2 file format
   // specification: https://numpy.org/neps/nep-0001-npy-format.html
   // while this allocation could be sizable, it is performed on the host.
-  auto block_start = align_down(offset, o_direct_alignm);
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/util/numpy.cc:205]: Mhh, based on the linked page we support v1 as the offset to the header is computed as `int64 offset = 6 + 1 + 1 + 2;` while for v2 it should be `int64 offset = 6 + 1 + 1 + 4;`.
Or it just works as with little endian 3rd and 4th byte of the header length are usually 0 and if we read them as a part of the header itself it is just NULL that is disregarded by further parsing.

File: dali/util/numpy.cc [hunk 5]
```diff
@@ -183,45 +183,50 @@ void ParseHeaderItself(HeaderData &parsed_header, char *data, size_t header_len)
...
+    nread = file->ReadAt(token_mem.get(), aligned_token_header_len, 0);
   } else {
     // restore overriden character
-    token[10] = char_tmp;
+    token[token_len] = char_tmp;
   }
-  token = token_mem.get() + alignment_offset(offset, o_direct_alignm);
-  DALI_ENFORCE(nread <= static_cast<Index>(aligned_len) &&
-               nread >= static_cast<Index>(std::min(src->Size(), aligned_len)),
+  DALI_ENFORCE(nread <= static_cast<Index>(aligned_token_header_len) &&
```
Comment (JanuszL) [inline dali/util/numpy.cc:222]: This should probably go to the `if` part.

Comment (stiepan) [reply dali/util/numpy.cc]: Right! Done.

**Comment by stiepan** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

File: dali/util/numpy.cc [hunk 6]
```diff
@@ -183,45 +183,51 @@ void ParseHeaderItself(HeaderData &parsed_header, char *data, size_t header_len)
...
   auto header_len = GetHeaderLen(token);
 
-  int64 offset = 6 + 1 + 1 + 2;
-  // the header_len can be 4GiB according to the NPYv2 file format
-  // specification: https://numpy.org/neps/nep-0001-npy-format.html
-  // while this allocation could be sizable, it is performed on the host.
-  auto block_start = align_down(offset, o_direct_alignm);
-  auto block_end = align_up(header_len + 1, o_direct_alignm);
-  aligned_len = align_up(block_end - block_start, o_direct_read_len_alignm);
+  // The header_len have up to 2**16 - 1 bytes. We do not support V2 headers
```
Comment (JanuszL) [inline dali/util/numpy.cc:206]: Shouldn't we enforce V1 in such a case? Or V2 can be also a plain numpy file without structured arrays.

**Review by klecki — APPROVED** [review]

File: dali/util/numpy.cc (hunk 6 above)
Comment (JanuszL) [reply dali/util/numpy.cc:206]: So how about adding a warning that if v2 is encountered some features may not work.

**Comment by stiepan** [comment]
!build

**Comment by stiepan** [comment]
!build

**Review by JanuszL — APPROVED** [review]

**Review by JanuszL — APPROVED** [review]

File: dali/pipeline/pipeline.cc [hunk 7]
```diff
@@ -162,15 +162,21 @@ Pipeline::~Pipeline() {
...
-        break;
-      case OpType::MIXED:
-        dynamic_cast<InputOperator<MixedBackend> *>(op_ptr)->BreakWaiting();
-        break;
-      case OpType::GPU:
-        dynamic_cast<InputOperator<GPUBackend> *>(op_ptr)->BreakWaiting();
-        break;
+      case OpType::CPU: {
+        auto *cpu_op_ptr = dynamic_cast<InputOperator<CPUBackend> *>(op_ptr);
+        assert(cpu_op_ptr);
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/pipeline/pipeline.cc:167]: Understood

**Pull request merged**